src/components/reports/cost.ts

```diff
@@ -0,0 +1,169 @@
+
+  const billingClient = new BillingClient({
+    apiEndpoint: ctx.app.billingAPI,
+    accessToken: ctx.token.accessToken,
+  });
+
+  const cf = new CloudFoundryClient({
+    accessToken: ctx.token.accessToken,
+    apiEndpoint: ctx.app.cloudFoundryAPI,
+  });
+
+  const orgs = await cf.organizations();
+  const orgGUIDs = orgs.map(o => o.metadata.guid);
+
+  const orgQuotas: {[key: string]: IOrganizationQuota} = (await Promise.all(
+    orgs.map(async (org: IOrganization) => {
+      const quota = await cf.organizationQuota(org.entity.quota_definition_guid);
+      return {[org.metadata.guid]: quota};
+    }),
+  ))
+  .reduce(/* istanbul ignore next */ (accumulator, quota) => {
+    return { ...accumulator, ...quota };
+  });
+
+  const billableEvents = await billingClient.getBillableEvents({
+    rangeStart, rangeStop, orgGUIDs,
+  });
+
+  const orgBillableEvents = aggregateBillingEvents(billableEvents);
+  const orgCostRecords = createOrgCostRecords(
+    orgs,
+    orgQuotas,
+    orgBillableEvents,
+  );
+  const quotaCostRecords = createQuotaCostRecords(orgCostRecords);
+
+  const totalBillables = sumRecords(billableEvents);
+  const billableEventCount = billableEvents.length;
+
+  return {
+    body: costReportTemplate.render({
+      date: moment(rangeStart).format('MMMM YYYY'),
+      orgCostRecords: Object.values(orgCostRecords),
+      quotaCostRecords,
+      totalBillables,
+      billableEventCount,
+    }),
+  };
+}
+
+export function aggregateBillingEvents(
+  billableEvents: ReadonlyArray<IBillableEvent>,
+): {[key: string]: ReadonlyArray<IBillableEvent>} {
+  return billableEvents
+    .reduce(
+      (
+        accumulator: {[key: string]: ReadonlyArray<IBillableEvent>},
+        billableEvent: IBillableEvent,
+      ) => {
+        const accEvents = accumulator[billableEvent.orgGUID] || [];
+        return {
+          ...accumulator,
+          [billableEvent.orgGUID]: accEvents.concat([billableEvent]),
+        };
+      }, {});
+}
+
+export function createOrgCostRecords(
+  orgs: ReadonlyArray<IOrganization>,
+  orgQuotas: {[key: string]: IOrganizationQuota},
+  orgBillableEvents: {[key: string]: ReadonlyArray<IBillableEvent>},
+): ReadonlyArray<IOrgCostRecord> {
+  return orgs
+  .map(org => {
+    const quota = orgQuotas[org.metadata.guid];
+    const billableEvents = orgBillableEvents[org.metadata.guid] || [];
+
+    const incVAT = billableEvents.reduce(
+      (acc: number, e: IBillableEvent) => acc + e.price.incVAT, 0,
+    );
+    const exVAT = billableEvents.reduce(
+      (acc: number, e: IBillableEvent) => acc + e.price.exVAT, 0,
+    );
+
+    return {
+      orgGUID: org.metadata.guid,
+      orgName: org.entity.name,
+      quotaGUID: quota.metadata.guid,
+      quotaName: quota.entity.name,
+      incVAT, exVAT,
+    };
+  });
+}
+
+export function createQuotaCostRecords(
+  records: ReadonlyArray<IOrgCostRecord>,
+): ReadonlyArray<IQuotaCostRecord> {
+  const costsByQuota = records
+  .reduce((
```

Comment:
I wonder if it is worth it to iterate across all events and aggregate all the cost for non existent orgs and quotas

Reply:
Interesting. Due to the way you have to request events for a specific org, I would hope that there would be no "left-over" events... and so a change to paas-billing would need to be made for that.